src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -3,48 +3,30 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
-using System.Collections.Generic;
-using System.Diagnostics;
 using System.IO;
-using System.Net;
-using System.Net.Sockets;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.Data.SqlClient;
-using Microsoft.Data.SqlClient.SNI;
 
 namespace Microsoft.Data.SqlClientX.Handlers.TransportCreation
 {
-    internal sealed class TransportCreationHandler : IHandler<ConnectionHandlerContext>
+    internal sealed class TransportCreationHandler : ContextHandler<ConnectionHandlerContext>
     {
-        private const int KeepAliveIntervalSeconds = 1;
-        private const int KeepAliveTimeSeconds = 30;
+        private readonly ReturningHandler<ConnectionHandlerContext, Stream> _streamCreationChain;
 
-        #if NET8_0_OR_GREATER
-        private static readonly TimeSpan DefaultPollTimeout = TimeSpan.FromSeconds(30);
-        #else
-        private const int DefaultPollTimeout = 30 * 100000; // 30 seconds as microseconds
-        #endif
-
-        /// <inheritdoc />
-        public IHandler<ConnectionHandlerContext> NextHandler { get; set; }
+        public TransportCreationHandler()
+        {
+            // Construct the chain of responsibility for handling the connections
+            _streamCreationChain = new SharedMemoryTransportCreationHandler();
+            _streamCreationChain.NextHandler = new TcpTransportCreationHandler();
+            _streamCreationChain.NextHandler.NextHandler = new SharedMemoryTransportCreationHandler();
+        }
 
         /// <inheritdoc />
-        public async ValueTask Handle(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        public override async ValueTask Handle(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
         {
-            Debug.Assert(context.DataSource is not null, "context.DataSource is null");
-
             try
             {
-                // @TODO: Build CoR for handling the different protocols in order
-                if (context.DataSource.ResolvedProtocol is DataSource.Protocol.TCP)
-                {
-                    context.ConnectionStream = await HandleTcpRequest(context, isAsync, ct).ConfigureAwait(false);
-                }
-                else
-                {
-                    throw new NotImplementedException();
-                }
+                context.ConnectionStream = await _streamCreationChain.Handle(context, isAsync, ct);
```

Comment:
The usual chain in connectivity is SM > TCP > NP in the absence of a protocol.

If only NP is available on the server, then SM cannot connect, neither can TCP. So what will happen ? These two will throw? and try their next handler? 
What if there is no server? Then SM/TCP/NP all will fail, but then what will NP handler do? It will swallow the exception and pass the control to HandleNext?

Reply:
For question 1) Right, SM will catch exception and pass onto the TCP handler. TCP handler catches exception and passes onto the NP handler. NP handler would succeed and return the stream.

For question 2) Ok, that's a good point. It's currently implemented to try SM, fail and pass to next, try TCP, fail and pass to next, try NP, fail, then pass to next, but there isn't any next so it throws no suitable handler found. Since we have an enum here, we could presume that the getting that exception means we couldn't connect at all, but I'm not sure. What kind of exception would you want in that situation?

Reply:
Ok, I have an improved implementation (probably overengineered to heck) it will be pushed shortly.